### PR TITLE
Verify class compatibility before attempting cast

### PR DESCRIPTION
**Pull request by breedx-splk**
There was a case in the field on an IBM J9 JVM, where an instance returned from `ManagementFactory.getThreadMXBean()` could not be cast safely to the `com.sun.management.ThreadMXBean`, even thought that class can be found on the classpath.

It looks like [other implementations](https://github.com/elastic/apm-agent-java/issues/505) have also encountered this, and so this fix is inspired by those [other workarounds](https://github.com/felixbarny/apm-agent-java/commit/bb7f7712ca7c4126aeb3d22709777ba7684ab14a).

### CODE REVIEW

File: instrumentation/jvm-metrics/src/main/java/com/splunk/opentelemetry/instrumentation/jvmmetrics/micrometer/MicrometerAllocatedMemoryMetrics.java [hunk 1]
```diff
@@ -19,6 +19,7 @@
 import static com.splunk.opentelemetry.instrumentation.jvmmetrics.AllocatedMemoryMetrics.METRIC_NAME;
 
 import com.splunk.opentelemetry.instrumentation.jvmmetrics.AllocatedMemoryMetrics;
+import io.micrometer.common.lang.NonNull;
```
Comment (laurit) [inline instrumentation/jvm-metrics/src/main/java/com/splunk/opentelemetry/instrumentation/jvmmetrics/micrometer/MicrometerAllocatedMemoryMetrics.java:22]: wrong import?

File: instrumentation/jvm-metrics/src/main/java/com/splunk/opentelemetry/instrumentation/jvmmetrics/AllocatedMemoryMetrics.java [hunk 2]
```diff
@@ -53,13 +52,22 @@ private static boolean hasComSunThreadMXBean() {
...
+      Class<?> mxBeanClass = Class.forName("com.sun.management.ThreadMXBean");
+      return mxBeanClass.isInstance(ManagementFactory.getThreadMXBean());
+    } catch (Exception e) {
+      return false;
+    }
+  }
+
   private static boolean isThreadAllocatedMemoryEnabled() {
-    ThreadMXBean threadBean = (ThreadMXBean) ManagementFactory.getThreadMXBean();
+    ThreadMXBean threadBean = (com.sun.management.ThreadMXBean) ManagementFactory.getThreadMXBean();
```
Comment (laurit) [inline instrumentation/jvm-metrics/src/main/java/com/splunk/opentelemetry/instrumentation/jvmmetrics/AllocatedMemoryMetrics.java:65]: unnecessary qualified class name in the cast

File: instrumentation/jvm-metrics/src/main/java/com/splunk/opentelemetry/instrumentation/jvmmetrics/AllocatedMemoryMetrics.java [hunk 3]
```diff
@@ -53,13 +52,22 @@ private static boolean hasComSunThreadMXBean() {
     }
   }
 
+  private static boolean mxBeanTypeIsCompatible() {
+    try {
+      Class<?> mxBeanClass = Class.forName("com.sun.management.ThreadMXBean");
```
Comment (laurit) [inline instrumentation/jvm-metrics/src/main/java/com/splunk/opentelemetry/instrumentation/jvmmetrics/AllocatedMemoryMetrics.java:57]: use `Class.forName("com.sun.management.ThreadMXBean", false, AllocatedMemoryMetrics.class.getClassLoader());` like in `hasComSunThreadMXBean()` or load the class in separate method and keep it in a static field or pass it to these methods

File: instrumentation/jvm-metrics/src/main/java/com/splunk/opentelemetry/instrumentation/jvmmetrics/AllocatedMemoryMetrics.java [hunk 4]
```diff
@@ -25,19 +25,18 @@
 
 public class AllocatedMemoryMetrics {
   public static final String METRIC_NAME = "process.runtime.jvm.memory.allocated";
-  private static final boolean hasComSunThreadMXBean = hasComSunThreadMXBean();
   private final AllocationTracker allocationTracker = createAllocationTracker();
 
-  public boolean isAvailable() {
-    return allocationTracker != null;
+  public boolean isUnavailable() {
```
Comment (laurit) [inline instrumentation/jvm-metrics/src/main/java/com/splunk/opentelemetry/instrumentation/jvmmetrics/AllocatedMemoryMetrics.java:30]: there are more usages of `isAvailable` in this module

File: instrumentation/jvm-metrics/src/main/java/com/splunk/opentelemetry/instrumentation/jvmmetrics/micrometer/MicrometerAllocatedMemoryMetrics.java (hunk 1 above)
Comment (breedx-splk) [reply instrumentation/jvm-metrics/src/main/java/com/splunk/opentelemetry/instrumentation/jvmmetrics/micrometer/MicrometerAllocatedMemoryMetrics.java:22]: Nope, micrometer has their own. 🙄

File: instrumentation/jvm-metrics/src/main/java/com/splunk/opentelemetry/instrumentation/jvmmetrics/AllocatedMemoryMetrics.java (hunk 2 above)
Comment (breedx-splk) [reply instrumentation/jvm-metrics/src/main/java/com/splunk/opentelemetry/instrumentation/jvmmetrics/AllocatedMemoryMetrics.java:65]: This was intentional. This code is already rather hard to follow because there are two classes called `ThreadMXBean` in different packages, and I thought that the fully qualified cast actually helped here.

File: instrumentation/jvm-metrics/src/main/java/com/splunk/opentelemetry/instrumentation/jvmmetrics/AllocatedMemoryMetrics.java (hunk 4 above)
Comment (breedx-splk) [reply instrumentation/jvm-metrics/src/main/java/com/splunk/opentelemetry/instrumentation/jvmmetrics/AllocatedMemoryMetrics.java:30]: I wasn't trying for completeness, just the task at hand. I guess I can change those other unrelated ones as well to be consistent.

File: instrumentation/jvm-metrics/src/main/java/com/splunk/opentelemetry/instrumentation/jvmmetrics/micrometer/MicrometerAllocatedMemoryMetrics.java (hunk 1 above)
Comment (laurit) [reply instrumentation/jvm-metrics/src/main/java/com/splunk/opentelemetry/instrumentation/jvmmetrics/micrometer/MicrometerAllocatedMemoryMetrics.java:22]: why on earth would you want to use the micrometer one?

File: .github/workflows/ci.yaml [hunk 5]
```diff
@@ -43,7 +43,7 @@ jobs:
           java-version: 17
 
       - name: Cache Gradle Wrapper
-        uses: actions/cache@v3.3.1
+        uses: actions/cache@v3.3.2
```
[the remark this answers is not in the document]
Comment (breedx-splk) [reply .github/workflows/ci.yaml:46]: had to merge due to conflict I think.

File: instrumentation/jvm-metrics/src/main/java/com/splunk/opentelemetry/instrumentation/jvmmetrics/micrometer/MicrometerAllocatedMemoryMetrics.java (hunk 1 above)
Comment (breedx-splk) [reply instrumentation/jvm-metrics/src/main/java/com/splunk/opentelemetry/instrumentation/jvmmetrics/micrometer/MicrometerAllocatedMemoryMetrics.java:22]: This class is an implementation of a micrometer interface. That interface uses the micrometer specific annotation. Using an annotation that doesn't match the actual interface declaration seems hacky.

**Comment by breedx-splk** [comment]
Superseded by #1435
